server/bin/pbench-restore-and-unpack-tarballs.py

```diff
@@ -0,0 +1,253 @@
+                    yield tb_dt, c_entry.name, tb
+
+
+def main(options):
+    if not options.cfg_name:
+        print(
+            f"{_NAME_}: ERROR: No config file specified; set"
+            " _PBENCH_SERVER_CONFIG env variable",
+            file=sys.stderr,
+        )
+        return 2
+
+    try:
+        config = pbench.PbenchConfig(options.cfg_name)
+    except pbench.BadConfig as e:
+        print(f"{_NAME_}: {e}", file=sys.stderr)
+        return 2
+
+    archive = config.ARCHIVE
+    archive_p = Path(archive).resolve(strict=True)
+
+    if not archive_p.is_dir():
+        print(
+            f"The configured ARCHIVE directory, {archive}, is not a valid directory",
+            file=sys.stderr,
+        )
+        return 2
+
+    backup = config.conf.get("pbench-server", "pbench-backup-dir")
+    backup_p = Path(backup).resolve(strict=True)
+
+    if not backup_p.is_dir():
+        print(
+            "The configured pbench-backup-dir directory, {backup}, is not a"
+            " valid directory",
+            file=sys.stderr,
+        )
+        return 2
+
+    start = pbench._time()
+
+    print(f"Restoring tar balls from backup volume, {backup} (started at: {start})")
+
+    gen = gen_list(backup_p)
+    tbs_sorted = sorted(
+        [
+            (tb_dt, controller_name, tar_ball_name)
+            for tb_dt, controller_name, tar_ball_name in gen
+        ],
+        reverse=True,
+    )
+    tbs_cnt = len(tbs_sorted)
+    tbs_left = tbs_cnt
+    tbs_existing = 0
+    tbs_restored = 0
+    for tb_dt, ctrl, tb in tbs_sorted:
+        try:
+            a_tb = archive_p / ctrl / tb.name
+            tbs_left -= 1
+
+            if a_tb.exists():
+                tbs_existing += 1
+                print(
+                    f"Exists {tbs_existing:d} ({tbs_restored:d} restored) of ({tbs_cnt:d}): {a_tb}",
+                    flush=True,
+                )
+                continue
+
+            # Create controller directory if it doesn't exist
+            ctrl_p = archive_p / ctrl
+            if not options.dry_run:
+                try:
+                    ctrl_p.mkdir()
+                except FileExistsError:
+                    pass
+                else:
+                    print(f"\nCreated controller directory, '{ctrl_p}'", flush=True)
+
+            # Copy tar ball to archive tree, along with its .md5 ...
+            cp_cmd = f"cp -a {tb} {tb}.md5 {ctrl_p}/"
+            print(f"\n{cp_cmd}", flush=True)
+            if not options.dry_run:
+                cp = subprocess.run(
+                    cp_cmd.replace('"', '\\"').replace("'", "\\'"),
+                    shell=True,
+                    stderr=subprocess.STDOUT,
+                )
+                if cp.returncode != 0:
+                    print(
+                        f"FAILURE: cp command: '{cp_cmd}': {cp.returncode}, {cp.stdout!r}",
+                        file=sys.stderr,
+                    )
+                    return 1
```

Comment:
The replacement at line 169 should presumably be done before the print at line 166, given that it will show up at line 175 if that appears.

Reply:
Ditto for lines 187, 181, and 194, respectively.